Properties instead of tokens, to allow for ran usages to change thing…

anton-larsen-persequor opened the pull request:
…s based on annotations of said tokens.

This change also should limit the number of token class instantiations

Commit:
Properties instead of tokens, to allow for ran usages to change things based on annotations of said tokens.

This change also should limit the number of token class instantiations

Commit:
Further optimized token usage and started caching tokens statically. The number of unique tokens in play for a given vm should not grow beyond where this is useful

emanuel-persequor commented on ran-core/src/main/java/io/ran/CrudRepoBaseQuery.java:
```diff
@@ -13,7 +13,7 @@
 	protected TypeDescriber<T> typeDescriber;
 	protected GenericFactory genericFactory;
-	protected Property currentProperty = Property.get(Token.of("t"), Clazz.of(Object.class));
+	protected Property currentProperty = null;
```
What is this field used for, it is not used by the class itself, since it's protected it could be used by subclasses... but it's not clear for what. Maybe a javacomment to explain, or remove it if it is not actually being used?

emanuel-persequor commented on ran-core/src/main/java/io/ran/MappingClassWriter.java:
```diff
@@ -17,20 +17,21 @@
 public class MappingClassWriter extends AutoMapperClassWriter {
 	Clazz mapperClazz;
-	public MappingClassWriter(Class clazz) {
+	public MappingClassWriter(Class clazz) throws NoSuchMethodException {
```
Why will the constructor throw this exception? Should it not handle it internally (as I guess it should not happen - but the java compiler doesn't know this).

emanuel-persequor commented on ran-core/src/main/java/io/ran/MappingClassWriter.java:
```diff
@@ -375,109 +393,123 @@ private void createHydrator() {
+					ce.load("clazz");
+					ce.load("this");
 					ce.getField(getSelf(), "_serializerFor" + field.getName(), Clazz.of(ISerializer.class));
 					{
```
Funny use of curly bracers - maybe a java comment to explain what each block does - would help readability (and allow future modifications more easily)

Commit:
Fixed a couple of issues in token caching and added a few code comments

Commit:
Don't require that collections fields are null as part of lazy loading check

Doing it this way  has the upside that you can build your models with the collection fields initialized as empty list, which would make the model behave similarly in unit tests and in integration tests (the collection elements wouldn't be null in either

andrepersequor commented on ran-core/src/main/java/io/ran/Property.java:
```diff
@@ -2,16 +2,12 @@
 
 public class Property<T> {
 	private Token token;
+	private String snakeCase;
```
why this? is easier to handle in the mapper generation code?

andrepersequor commented on ran-core/src/main/java/io/ran/Property.java:
```diff
@@ -183,13 +237,25 @@ public KeySet mapProperties(PropertyList other) {
 			try {
 				return getOptional(token).orElseThrow(() -> new RuntimeException("Could not find property by token: " + token.toString() + " on " + stream().findAny().map(p -> p.getOn().clazz.getName()).orElse("unknown class")));
-			} catch (NullPointerException e) {
+			} catch (Exception e) {
+				throw e;
```
this seems pointless

andrepersequor commented on ran-core/src/main/java/io/ran/Property.java:
```diff
@@ -149,7 +203,7 @@ public KeySets keys() {
 				}
 			}
 			if (keys.isEmpty() && contains(Token.of("id"))) {
```
this means that if no key annotation was added to the class, will try to retrieve as a key a field called id right?

andrepersequor commented on ran-core/src/main/java/io/ran/QueryClassWriter.java:
```diff
@@ -76,10 +76,11 @@ protected void build() {
 				MethodWriter mw = method(Access.of(m.getModifiers()), new MethodSignature(m));
 				mw.load(0);
-				mw.getField(getSelf(), "currentProperty", Clazz.of(Property.class));
+				mw.load(0);
```
why this is needed to be loaded twice, unlike before?

Commit:
Resolved issues with changing the name of indexed (primary or regular) columns using the DbName annotation, by ensuring we pass around the property when available

Commit:
add more missing fixes to use properties instead of tokens

andrepersequor reviewed (approved):
should be good to go
